### PR TITLE
nsqd: configurable HTTP client transport timeout

### LINKED ISSUE #680: nsq_to_http: improve --http-timeout
by default http-timeout(20s), on the same deadlineConn(api_request.go), the second request's timeout is less than 20s.
api_request.go:
```go
func (c *deadlinedConn) Read(b []byte) (n int, err error) {
	// it is not correct to set deadline on read operation, 
	// http connection doesn't support reading timeout
	// instead, we can set  http.Transport.ResponseHeaderTimeout.
	c.Conn.SetReadDeadline(time.Now().Add(c.Timeout))
	return c.Conn.Read(b)
}

func (c *deadlinedConn) Write(b []byte) (n int, err error) {
	// ok, it's right
	c.Conn.SetWriteDeadline(time.Now().Add(c.Timeout))
	return c.Conn.Write(b)
}

// A custom http.Transport with support for deadline timeouts
func NewDeadlineTransport(timeout time.Duration) *http.Transport {
	transport := &http.Transport{
		Dial: func(netw, addr string) (net.Conn, error) {
			c, err := net.DialTimeout(netw, addr, timeout)
			if err != nil {
				return nil, err
			}
			return &deadlinedConn{timeout, c}, nil
		},
		//  http reading timeout is here
		ResponseHeaderTimeout: timeout,
	}
	return transport
}
```

### LINKED ISSUE #715: nsqd: configurable HTTP client transport timeout
see https://github.com/nsqio/nsq/pull/683#issuecomment-155947122

### PR DESCRIPTION
**Pull request by kenjones-cisco**
Adds configuration options HTTPClientConnectTimeout and
HTTPClientRequestTimeout to control the connection and request timeout
repectively of the HTTP client.

Also added to the following app binaries:

- nsqadmin
- nsq_stat
- nsq_to_file
- nsq_to_http

Closes #715
Closes #680

### CODE REVIEW

**Comment by mreiferson** [comment]
@kenjones-cisco thanks for this!

I have a high level comment first.

If we're going to land this, we should probably just do #680 as well, because the timeout as implemented is really a read/write deadline on the connection and has unexpected behavior.

If we do #680, we should introduce two configurable values, `--http-client-connect-timeout` and `--http-client-request-timeout`. They will apply to the custom `Dial` and to the [HTTP client's](https://golang.org/pkg/net/http/#Client) `Timeout` property, respectively.

What do you think?

**Comment by kenjones-cisco** [comment]
@mreiferson That makes sense. Let me take a look.

**Comment by mreiferson** [comment]
Also, let's not forget to update `nsqd`'s example config file with these options.

**Comment by kenjones-cisco** [comment]
@mreiferson changes made, just saw the comment about the example config file. Based on #715 I thought it was command line options. For nsqd and nsqadmin, the options should be specified for the configuration file as well?

**Comment by mreiferson** [comment]
> For nsqd and nsqadmin, the options should be specified for the configuration file as well?

Might only be `nsqd` - it uses [`go-options`](https://github.com/mreiferson/go-options) to resolve configuration coming from multiple sources (the struct's defaults, command line flags, and config files).

File: internal/http_api/api_request.go [hunk 1]
```diff
@@ -19,7 +19,6 @@ type deadlinedConn struct {
 }
 
 func (c *deadlinedConn) Read(b []byte) (n int, err error) {
-	c.Conn.SetReadDeadline(time.Now().Add(c.Timeout))
```
Comment (mreiferson) [inline internal/http_api/api_request.go]: we should drop the one in `Write` too

File: internal/http_api/api_request.go [hunk 2]
```diff
@@ -38,6 +37,7 @@ func NewDeadlineTransport(timeout time.Duration) *http.Transport {
 			}
 			return &deadlinedConn{timeout, c}, nil
 		},
+		ResponseHeaderTimeout: requestTimeout,
```
Comment (mreiferson) [inline internal/http_api/api_request.go]: I don't _think_ we need to set this, considering we're bounding the _entire_ request by the same timeout.

File: apps/nsq_stat/nsq_stat.go [hunk 3]
```diff
@@ -27,6 +27,8 @@ var (
 	channel          = flag.String("channel", "", "NSQ channel")
 	statusEvery      = flag.Duration("status-every", -1, "(deprecated) duration of time between polling/printing output")
 	interval         = flag.Duration("interval", 2*time.Second, "duration of time between polling/printing output")
+	httpConnectTimeout = flag.Duration("http-client-connect-timeout", 2*time.Second, "timeout for HTTP connect")
+	httpRequestTimeout = flag.Duration("http-client-request-timeout", 2*time.Second, "timeout for HTTP read/write (each)")
```
Comment (mreiferson) [inline apps/nsq_stat/nsq_stat.go]: The description should be `timeout for HTTP request`, it's not the read/write anymore after this set of changes.

(repeat for all locations of this description)

**Comment by kenjones-cisco** [comment]
@mreiferson  Looks like the DiskQueue test failure is back; otherwise all changes based on feedback are completed.

File: apps/nsq_stat/nsq_stat.go [hunk 4]
```diff
@@ -27,6 +27,8 @@ var (
 	channel          = flag.String("channel", "", "NSQ channel")
 	statusEvery      = flag.Duration("status-every", -1, "(deprecated) duration of time between polling/printing output")
 	interval         = flag.Duration("interval", 2*time.Second, "duration of time between polling/printing output")
+	httpConnectTimeout = flag.Duration("http-client-connect-timeout", 2*time.Second, "timeout for HTTP connect")
+	httpRequestTimeout = flag.Duration("http-client-request-timeout", 2*time.Second, "timeout for HTTP request")
```
Comment (mreiferson) [inline apps/nsq_stat/nsq_stat.go]: Didn't think of this before, but now that we're setting a real request timeout, which encapsulates connection time, we should set the default for request timeout something greater than connect, say `5s`?

cc @jehiah

**Comment by kenjones-cisco** [comment]
@mreiferson Just let me know when a default is decided.

**Comment by mreiferson** [comment]
@kenjones-cisco let's go with `5s` for now, @jehiah can chime in if he wants.

**Comment by kenjones-cisco** [comment]
@mreiferson all done

File: apps/nsq_to_http/nsq_to_http.go [hunk 5]
```diff
@@ -42,9 +42,12 @@ var (
 	numPublishers = flag.Int("n", 100, "number of concurrent publishers")
 	mode          = flag.String("mode", "hostpool", "the upstream request mode options: multicast, round-robin, hostpool (default), epsilon-greedy")
 	sample        = flag.Float64("sample", 1.0, "% of messages to publish (float b/w 0 -> 1)")
-	httpTimeout   = flag.Duration("http-timeout", 20*time.Second, "timeout for HTTP connect/read/write (each)")
-	statusEvery   = flag.Int("status-every", 250, "the # of requests between logging status (per handler), 0 disables")
-	contentType   = flag.String("content-type", "application/octet-stream", "the Content-Type used for POST requests")
+	// deprecate? in favor of http-client-connect-timeout, http-client-request-timeout
```
Comment (mreiferson) [inline apps/nsq_to_http/nsq_to_http.go]: Good catch, can you put the `// TODO: ` prefix on this so it shows up in searches?

File: apps/nsq_stat/nsq_stat.go [hunk 6]
```diff
@@ -22,14 +22,16 @@ import (
...
-	interval         = flag.Duration("interval", 2*time.Second, "duration of time between polling/printing output")
-	countNum         = numValue{}
-	nsqdHTTPAddrs    = app.StringArray{}
-	lookupdHTTPAddrs = app.StringArray{}
+	showVersion        = flag.Bool("version", false, "print version")
+	topic              = flag.String("topic", "", "NSQ topic")
+	channel            = flag.String("channel", "", "NSQ channel")
+	statusEvery        = flag.Duration("status-every", -1, "(deprecated) duration of time between polling/printing output")
+	interval           = flag.Duration("interval", 2*time.Second, "duration of time between polling/printing output")
+	httpConnectTimeout = flag.Duration("http-client-connect-timeout", 5*time.Second, "timeout for HTTP connect")
```
Comment (mreiferson) [inline apps/nsq_stat/nsq_stat.go]: I meant for the connect timeout to remain at `2s` and bump request timeout to `5s`. It doesn't really make sense for them to be equal to each other.

File: contrib/nsqd.cfg.example [hunk 7]
```diff
@@ -21,6 +21,11 @@ nsqlookupd_tcp_addresses = [
     "127.0.0.1:4160"
 ]
 
+## duration to wait before HTTP client connection timeout
+http_client_connect_timeout = "2s"
+
+## duration to wait before HTTP client request timeout
+http_client_request_timeout = "2s"
```
Comment (mreiferson) [inline contrib/nsqd.cfg.example]: missed a `5s` bump

File: internal/http_api/api_request.go [hunk 8]
```diff
@@ -46,12 +44,13 @@ type Client struct {
 	c *http.Client
 }
 
-func NewClient(tlsConfig *tls.Config) *Client {
-	transport := NewDeadlineTransport(2 * time.Second)
+func NewClient(tlsConfig *tls.Config, clientTimeout time.Duration, requestTimeout time.Duration) *Client {
```
Comment (mreiferson) [inline internal/http_api/api_request.go]: `s/clientTimeout/connectTimeout`

File: internal/http_api/api_request.go [hunk 9]
```diff
@@ -19,12 +19,10 @@ type deadlinedConn struct {
 }
 
 func (c *deadlinedConn) Read(b []byte) (n int, err error) {
-	c.Conn.SetReadDeadline(time.Now().Add(c.Timeout))
 	return c.Conn.Read(b)
 }
 
 func (c *deadlinedConn) Write(b []byte) (n int, err error) {
-	c.Conn.SetWriteDeadline(time.Now().Add(c.Timeout))
```
Comment (mreiferson) [inline internal/http_api/api_request.go]: Now that we've removed these, we need to set `ResponseHeaderTimeout` to our configured request timeout on the `http.Transport` instantiation below on line 30. `NewDeadlineTransport` will need to take both connect timeout and request timeout.

Comment (mreiferson) [inline internal/http_api/api_request.go]: (otherwise `DeadlineTransport` won't actually have request timeouts)

Comment (kenjones-cisco) [inline internal/http_api/api_request.go]: I had that previously, and your previous comments said it shouldn't be needed as entire request is now being bounded.

Comment (mreiferson) [inline internal/http_api/api_request.go]: Right, sorry about the back and forth. What I realized is that, when `DeadlineTransport` is used for proxying (`nsqadmin`) there is no other request timeout being applied (unlike when it's used with an HTTP client).

We could set the `ResponseHeaderTimeout` only in the case where it's used for proxying?

**Comment by kenjones-cisco** [comment]
@mreiferson Let me know if there is any more feedback.

File: apps/nsq_to_file/nsq_to_file.go [hunk 10]
```diff
@@ -45,6 +45,9 @@ var (
 	rotateSize     = flag.Int64("rotate-size", 0, "rotate the file when it grows bigger than `rotate-size` bytes")
 	rotateInterval = flag.Duration("rotate-interval", 0*time.Second, "rotate the file every duration")
 
+	httpConnectTimeout = flag.Duration("http-client-connect-timeout", 5*time.Second, "timeout for HTTP connect")
```
Comment (mreiferson) [inline apps/nsq_to_file/nsq_to_file.go]: should be `2s`

File: apps/nsq_to_http/nsq_to_http.go [hunk 11]
```diff
@@ -42,9 +42,12 @@ var (
 	numPublishers = flag.Int("n", 100, "number of concurrent publishers")
 	mode          = flag.String("mode", "hostpool", "the upstream request mode options: multicast, round-robin, hostpool (default), epsilon-greedy")
 	sample        = flag.Float64("sample", 1.0, "% of messages to publish (float b/w 0 -> 1)")
-	httpTimeout   = flag.Duration("http-timeout", 20*time.Second, "timeout for HTTP connect/read/write (each)")
-	statusEvery   = flag.Int("status-every", 250, "the # of requests between logging status (per handler), 0 disables")
-	contentType   = flag.String("content-type", "application/octet-stream", "the Content-Type used for POST requests")
+	// TODO: remove; deprecated in favor of http-client-connect-timeout, http-client-request-timeout
+	httpTimeout        = flag.Duration("http-timeout", 20*time.Second, "timeout for HTTP connect/read/write (each)")
+	httpConnectTimeout = flag.Duration("http-client-connect-timeout", 20*time.Second, "timeout for HTTP connect")
```
Comment (mreiferson) [inline apps/nsq_to_http/nsq_to_http.go]: should be 2s

File: nsqadmin/http.go [hunk 12]
```diff
@@ -83,7 +84,7 @@ func NewHTTPServer(ctx *Context) *httpServer {
 	router.Handle("GET", "/static/:asset", http_api.Decorate(s.staticAssetHandler, log, http_api.PlainText))
 	router.Handle("GET", "/fonts/:asset", http_api.Decorate(s.staticAssetHandler, log, http_api.PlainText))
 	if s.ctx.nsqadmin.getOpts().ProxyGraphite {
-		proxy := NewSingleHostReverseProxy(ctx.nsqadmin.graphiteURL, 20*time.Second)
+		proxy := NewSingleHostReverseProxy(ctx.nsqadmin.graphiteURL, 20*time.Second, 20*time.Second)
```
Comment (mreiferson) [inline nsqadmin/http.go]: these should use options?

File: nsqadmin/nsqadmin.go [hunk 13]
```diff
@@ -143,7 +143,7 @@ func (n *NSQAdmin) handleAdminActions() {
 		if err != nil {
 			n.logf("ERROR: failed to serialize admin action - %s", err)
 		}
-		httpclient := &http.Client{Transport: http_api.NewDeadlineTransport(10 * time.Second)}
+		httpclient := &http.Client{Transport: http_api.NewDeadlineTransport(10*time.Second, 10*time.Second)}
```
Comment (mreiferson) [inline nsqadmin/nsqadmin.go]: same here, use options?

File: nsqlookupd/nsqlookupd_test.go [hunk 14]
```diff
@@ -15,6 +15,11 @@ import (
 	"github.com/nsqio/nsq/internal/http_api"
 )
 
+const (
+	ConnectTimeout = 5 * time.Second
```
Comment (mreiferson) [inline nsqlookupd/nsqlookupd_test.go]: minor, but 2s for consistency

**Comment by mreiferson** [comment]
@kenjones-cisco found a few more inconsistencies, thanks for your patience!

**Comment by kenjones-cisco** [comment]
Some were previously hard-coded to a specific value so I left those to avoid breaking existing functionality. But I can make the changes.

**Comment by mreiferson** [comment]
> Some were previously hard-coded to a specific value so I left those to avoid breaking existing functionality. But I can make the changes.

@kenjones-cisco appreciate that, and I think we can still accomplish that by preserving those previously hard-coded values as the defaults for the flags, for those applications. What will be new (and potentially unexpected) is that we are now introducing a separate connection timeout in those  cases. I believe `2s` is a reasonable default.

**Comment by kenjones-cisco** [comment]
@mreiferson all changes completed

**Comment by mreiferson** [comment]
LGTM, thanks!

**Pull request merged**